### ISSUE TITLE
Popups should expose their accessibility tree

### ISSUE THREAD

**Issue by DataTriny** [slint-ui/slint#7209]
### Bug Description

When a popup element is shown, its accessibility tree doesn't seem to be merged with the parent one. I would expect that the root element of the popup would be added as a child of the element which contains the popup.

This is not even a case of keyboard focus not moving properly, the child elements of any popup are not part of the accessibility tree, even if they are configured to be.

### Reproducible Code (if applicable)

```slint
The issue can easily be reproduced with the todo example:

1. Open Accessibility Insights for Windows
2. Run the todo example: `cargo run -p todo`
3. Click on the Close button of the example window to make the confirmation popup appear
4. Observe in Accessibility Insights that the content of the popup cannot be found
```

### Environment Details

- Slint Version: commit d6f83a2af663de2128104fe4241e8b12f308db8d
- Platform/OS: Windows 11, Windows 10
- Programming Language: Rust
- Backend/Renderer: winit

### Product Impact

I'm working on supporting combo boxes in AccessKit. Not being able to interact with the list of possible choices that appear when expanding the combo box is a blocker.

**Comment by tronical** [comment]
Makes sense. I wonder how to do that with AccessKit for the case where the popup is not a top-level window (with winit).

**Comment by DataTriny** [comment]
It depends on the content of the popup: In the case of the todo example, since the popup is used to show a modal dialog, it should add a new element of role `Role::Dialog` as a child of the element containing the popup. As for combo boxes, it should add an element of type `Role::ListBox` (I'm working on this part but I can't see my changes due to this bug).

I can see that when the focus is requested inside a popup, the AccessKit adapter is notified about this but no tree update gets generated since all elements appear to have `AccessibilityRole::None`. I'm not even entirely sure which element actually gets the focus.

**Comment by tronical** [comment]
Ahhh I see. This makes sense but will require some additional work. The popup is not part of the same item tree (and accesskit tree), but we may need a way to make the association of the item tree of the popup belonging to a certain accessible element, so that we can build the AccessKit tree for this tree. Maybe an `accessible-popup: the_popup_name;` type of property.

Last resort is that we have a call/discussion in January in our public mattermost.

**Comment by DataTriny** [comment]
We have the `Node::owns` property which can be used to indicate that a node has children which are not semantically its descendents. But this could only work if the popup tree is contained in the main window tree. Do you mean this isn't the case currently? I cannot actually find the popup item in the descendents of the root window, but I only performed quick debugging. I'm not even sure if the root window tree is still there while a popup is open.

**Comment by tronical** [comment]
Thanks for the explanations, I think I understand this problem now. I've made https://github.com/slint-ui/slint/pull/7265 as initial plumbing, which I think can close this issue.

The next steps from here would be - as you've pointed out as well - support for the different types of popups. One aspect that comes to my mind that might need changing in the future is how exactly the accessibility tree is built. Take this markup for example:

```slint
export component Test inherits Window {
    Button {
        text: "Select Time";
        clicked => { time-picker.show(); }
    }
    time-picker := TimePickerPopup { ... }
}
```

and `TimePickerPopup` is - in the different styles - declared like this (pseudo code):

```slint
export component TimePickerPopup inherits PopupWindow {
    background-layer := MenuBorder { ... }
    dialog := Dialog {
    }
}
```

Now the question is: Where should `accessible-role: dialog;` be declared?

Right now it would have to be `PopupWindow`, as that becomes directly a child of `Tests`'s `Window`. But perhaps it would be much better to say that `Dialog` automatically implies `accessible-role: dialog;` - this needs to be done inside the slint-compiler. If there's agreement to that, then we instead of making the `PopupWindow` the access kit child, we could say that all accessible children (or the first accessible child?) of `PopupWindow` become the children instead. Then this would work automatically.

@ogoffart do you have an opinion on this?

**Comment by DataTriny** [comment]
`PopupWindow` components cannot be represented as is in the AccessKit tree as they currently get `Role::Window`. This role is reserved for native windows. If showing a popup can never spawn a native window, then I don't really see the point of keeping `PopupWindow` components in the accessibility tree. If it is more practical to still group all children of a popup window into a single node, `Role::GenericContainer` would probably make sense (it is automatically filtered out by AccessKit and is purely there for your convenience).

> But perhaps it would be much better to say that Dialog automatically implies accessible-role: dialog;

This is probably not flexible enough. In the case of the todo example, the proper role would actually be alert-dialog. I don't think the Slint compiler has enough information to figure out which role to use, even for a dialog, so it might be better to let app authors decide.

**Comment by tronical** [comment]
That makes a of sense. GenericContainer is the way to go then.
